component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/dao/Impl/TOTPSecretKeyDAOImpl.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.application.authenticator.totp.dao.Impl;
+
+import org.wso2.carbon.identity.application.authenticator.totp.dao.TOTPSecretKeyDAO;
+import org.wso2.carbon.identity.application.authenticator.totp.exception.TOTPException;
+import org.wso2.carbon.identity.application.authenticator.totp.util.SQLQueries;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+/**
+ * Default implementation of {@link TOTPSecretKeyDAO}. This handles DB operations related to configuring TOTP
+ * for the federated users.
+ */
+public class TOTPSecretKeyDAOImpl implements TOTPSecretKeyDAO {
+
+    @Override
+    public void setTOTPSecretKeyOfFederatedUser(String userId, String secretKey) throws TOTPException {
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection()) {
+            try (PreparedStatement preparedStatement = connection
+                    .prepareStatement(SQLQueries.SQL_INSERT_TOTP_SECRET_KEY)) {
+                preparedStatement.setString(1, userId);
+                preparedStatement.setString(2, secretKey);
+                preparedStatement.executeUpdate();
+                IdentityDatabaseUtil.commitTransaction(connection);
+            } catch (SQLException e1) {
+                IdentityDatabaseUtil.rollbackTransaction(connection);
+                throw new TOTPException("Error when store secret key for the federated user: " + userId, e1);
+            }
+        } catch (SQLException e) {
+            throw new TOTPException("Error while storing secret key against federated user id: " + userId, e);
```

Comment:
don't we need to rollback the transaction here?

Reply:
The transaction has been rollbacked from here[1].
[1] https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/94/files/d4c6fbbbca44d182a0895a5d01325f811d54e1cd#diff-282391ae961a984b218d10deb2f753ca2364a1ac59356fe683cfccf632637c84R47